lib/auth/export.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"encoding/pem"
+	"strings"
+	"time"
+
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/sshutils"
+	"github.com/gravitational/trace"
+)
+
+// ExportAuthoritiesRequest has the required fields to create an export authorities request.
+// An empty AuthType exports all types.
+type ExportAuthoritiesRequest struct {
+	AuthType                   string
+	Client                     ClientI
+	ExportAuthorityFingerprint string
+	ExportPrivateKeys          bool
+	CompatVersion              string
+}
+
+// ExportAuthorities returns the list of authorities in OpenSSH compatible formats
+// If the ExportAuthoritiesRequest.AuthType is present only prints keys for CAs of this type,
+// otherwise returns all keys concatenated
+func ExportAuthorities(ctx context.Context, req ExportAuthoritiesRequest) (string, error) {
+	ret := strings.Builder{}
+
+	var typesToExport []types.CertAuthType
+
+	// this means to export TLS authority
+	switch req.AuthType {
+	// "tls" is supported for backwards compatibility.
+	// "tls-host" and "tls-user" were added later to allow export of the user
+	// TLS CA.
+	case "tls", "tls-host":
+		return exportTLSAuthority(ctx, req.Client, types.HostCA, false, req.ExportPrivateKeys)
+	case "tls-user":
+		return exportTLSAuthority(ctx, req.Client, types.UserCA, false, req.ExportPrivateKeys)
+	case "db":
+		return exportTLSAuthority(ctx, req.Client, types.DatabaseCA, false, req.ExportPrivateKeys)
+	case "tls-user-der", "windows":
+		return exportTLSAuthority(ctx, req.Client, types.UserCA, true, req.ExportPrivateKeys)
+	}
+
+	// if no --type flag is given, export HostCA and UserCA.
```

Comment:
nit: I find it strange talking about client flags this deep into server code. Maybe:

```suggestion
	// If no AuthType is given, export both HostCA and UserCA.
```